src/archivematicaCommon/lib/elasticSearchFunctions.py

```diff
@@ -824,13 +841,131 @@ def _remove_tool_output_from_mets(doc):
+    and custom (non dublin core), both set through the
+    metadata/metadata.csv file.
+
+    These types are parsed and combined into a single dictionary of
+    metadata attributes.
+    """
+    result = {}
+    elements_with_metadata = []
+    for DMDID in file_pointer_division.attrib.get("DMDID", "").split():
+        dmdSec = ns.xml_find_premis(doc, 'mets:dmdSec[@ID="{}"]'.format(DMDID))
+        if dmdSec is not None:
+            elements_with_metadata += get_descriptive_section_metadata(dmdSec)
+    if elements_with_metadata:
+        result = combine_elements(elements_with_metadata)
+    return result
+
+
+def get_directory_metadata(directory, doc):
+    """Get descriptive or administrive metadata for a directory.
+
+    There are three types of metadata elements extracted:
+
+    1. Dublin core, set through the metadata/metadata.csv file or the
+       transfer/ingest metadata form in the dashboard
+    2. Custom (non dublin core), set through the metadata/metadata.csv
+       file
+    3. Bag/disk image metadata, set through the bag-info.txt file or
+       the disk image metadata form in the dashboard
+
+    These types are parsed and combined into a single dictionary of
+    metadata attributes. A marker element with the label of the
+    Directory entry is added to the result.
+    """
+    result = {}
+    elements_with_metadata = []
+    for DMDID in directory.attrib.get("DMDID", "").split():
+        dmdSec = ns.xml_find_premis(doc, 'mets:dmdSec[@ID="{}"]'.format(DMDID))
+        if dmdSec is not None:
+            elements_with_metadata += get_descriptive_section_metadata(dmdSec)
+    for ADMID in directory.attrib.get("ADMID", "").split():
+        amdSec = ns.xml_find_premis(doc, 'mets:amdSec[@ID="{}"]'.format(ADMID))
+        if amdSec is not None:
+            # look for bag/disk image metadata
+            elements_with_metadata += ns.xml_findall_premis(
+                amdSec, "mets:sourceMD/mets:mdWrap/mets:xmlData/transfer_metadata"
+            )
+    if elements_with_metadata:
+        # add an attribute with the relative path of the Directory entry
+        elements_with_metadata.append(get_relative_path_element(directory))
+        result = combine_elements(elements_with_metadata)
+    return result
+
+
+def _extract_aip_metadata(doc):
+    """Extract AIP metadata.
+
+    Given a doc representing a METS file, look for Directory entries
+    that have descriptive or administrative metadata in the physical
+    structMap and return dictionaries with metadata attributes for
+    each directory.
+    """
+    result = []
+    physical_struct_map = ns.xml_find_premis(doc, 'mets:structMap[@TYPE="physical"]')
+    if physical_struct_map is not None:
+        for directory in get_directories_with_metadata(physical_struct_map):
+            directory_metadata = get_directory_metadata(directory, doc)
+            if directory_metadata:
+                result.append(directory_metadata)
+    return result
```

Comment:
I'm wondering if this would be a good case for generators. Say you're looking at 10k directories, would generators help noticeably? If it's not much of an effort (yields and yields), perhaps we should?

Reply:
Thanks for the idea. `_get_directories_with_metadata` can be converted to a generator :+1:

Reply:
`_get_directories_with_metadata` ended as a single call to lxml's xpath.